Implement support for Hetzner DNS Cloud API

lbrocke opened the pull request:
This PR implements the new [Hetzner DNS Cloud API](https://docs.hetzner.cloud/reference/cloud#dns) ([released today](https://docs.hetzner.cloud/whats-new#2025-10-07-dns-in-hetzner-console) as beta), which will gradually replace the now deprecated Hetzner DNS Console API.

To support both APIs simultaneously, this commit bumps the package version to v2, more specifically `github.com/libdns/hetzner/v2`. The implementation is based on hcloud-go, Hetzner Cloud's official Go API client. I saw other libdns plugins being based on SDKs, so I thought this is okay as it reduces maintenance work.

I implemented this for usage in the [dns.providers.hetzner](https://github.com/caddy-dns/hetzner) caddy module and would appreciate a review. Please let me know what to change.

lbrocke committed:
Implement support for Hetzner DNS Cloud API

This commit implements the new Hetzner DNS Cloud API, which will gradually
replace the now deprecated Hetzner DNS Console API.

To support both APIs simultaneously, this commit bumps the package version to
v2, more specifically `github.com/libdns/hetzner/v2`. The implementation is
based on hcloud-go, Hetzner Cloud's official Go API client.

jooola commented:
Closes #12

cutlery commented:
+1.  Cheers!

jooola commented on provider.go:
```diff
@@ -1,184 +1,257 @@
+			// Ignore HTTP 404 errors.
+			if !errors.As(err, &hcloudErr) || hcloudErr.Code != hcloud.ErrorCodeNotFound {
+				return nil, err
+			}
```
```suggestion
			// Ignore HTTP 404 errors.
			if !hcloud.IsError(err, hcloud.ErrorCodeNotFound) {
				return nil, err
			}
```

jooola commented on provider.go:
```diff
@@ -1,184 +1,257 @@
-				Value: rr.Data,
-				TTL:   int(rr.TTL.Seconds()),
-			})
+	// Delete each possibly existing RRset, then re-create if with the appropriate records. Re-creating the RRset allows
+	// setting the TTL in the same API call; otherwise a SetRRSetRecords followed by ChangeRRSetTTL would be necessary.
```
Do we care about having atomic changes?

lbrocke replied:
Using `SetRRSetRecords` followed by `ChangeRRSetTTL` also wouldn't be atomic, and I think there is no other API call that can set both records and TTL at the same time. Do you have another idea?

Kugelschieber replied:
I don't think this matters tbh.

jooola replied:
Indeed, for ACME TXT record this might not be super important.

If we only handle ACME TXT records, I would argue we could spare ourselves few API requests by using `AddRecords` directly, and if the request failed because of conflicting TTL, retry after the rrset is deleted.

lbrocke replied:
> If we only handle ACME TXT records, I would argue we could spare ourselves few API requests by using `AddRecords` directly, and if the request failed because of conflicting TTL, retry after the rrset is deleted.

The `RecordSetter` interface requires `SetRecords` to delete existing records with the same name & type if not passed into the function:

> // For any (name, type) pair in the input, SetRecords ensures that the only
> // records in the output zone with that (name, type) pair are those that were
> // provided in the input.
> //
> // In RFC 9499 terms, SetRecords appends, modifies, or deletes records in the
> // zone so that for each RRset in the input, the records provided in the input
> // are the only members of their RRset in the output zone.
>
> https://github.com/libdns/libdns/blob/v1.1.1/libdns.go#L124

So I don't think we can use `AddRecords` here, as it appends instead of replaces records. We'd still need to run `DeleteRRSet` per RRSet before to ensure previously existing records are not present anymore.

jooola commented on util.go:
```diff
@@ -1,10 +1,75 @@
+func fromRecord(zone string, record libdns.Record) *hcloud.ZoneRRSet {
+	switch t := record.(type) {
+	case libdns.TXT:
+		// Make sure TXT record data is enclosed in quotation marks.
+		if !strings.HasPrefix(t.Text, "\"") || !strings.HasSuffix(t.Text, "\"") {
+			t.Text = fmt.Sprintf("\"%s\"", t.Text)
+		}
```
How do we handle longer TXT records split into multiple chunks of 255 chars?

lbrocke replied:
The plugin currently doesn't, and relies on the input to be correctly formatted. It's not a problem for `_acme-challenge` records, as their values are shorter than 255 characters, but adding other TXT records could result in API errors.

I can't find any documentation about how plugins should handle this in libdns. Would you like me to add this to be on the safe side?

cc @Kugelschieber

Kugelschieber replied:
This lib just needs to add acme records, so I don't think it's relevant here.

lbrocke replied:
Alright, I'll mark this thread as resolved then. If that becomes a problem in the future, I can add some splitting logic.

jooola reviewed (commented):
A few remarks:
- Why not wait for the returned actions?
- Are we responsible for converting to punycode or is that handled elsewhere?

Kugelschieber commented on README.md:
```diff
@@ -2,11 +2,14 @@
 [![godoc reference](https://img.shields.io/badge/godoc-reference-blue.svg)](https://pkg.go.dev/github.com/libdns/hetzner)
 
-This package implements the libdns interfaces for the [Hetzner DNS API](https://dns.hetzner.com/api-docs).
+This package implements the [libdns](https://github.com/libdns/libdns) interfaces for the Hetzner DNS APIs.
```
Add link to new docs: https://docs.hetzner.cloud/reference/cloud#dns

lbrocke replied:
The new docs are linked to in the second bullet point, does that suffice?

Kugelschieber replied:
Sorry, I haven't seen it. That should be sufficient :+1:

Kugelschieber commented on provider.go:
```diff
@@ -1,184 +1,257 @@
 }
 
-// New returns a new libdns provider for Hetzner.
-func New(token string) *Provider {
```
Keep the New constructor. This is very common in Go libs and provides a way to ask for all required fields instead of relying on the user to understand which fields are actually required. Not so much in this case of course, but it's still a good practice imo.

Kugelschieber commented on provider.go:
```diff
@@ -1,184 +1,257 @@
-	return &Provider{
-		AuthAPIToken: token,
-	}
+// getClient returns a hcloud.ZoneClient configured with the Provider's APIToken. The client is bootstrapped in the
```
Providing docs for unexported methods isn't really required, but we can keep it.

Kugelschieber commented on provider.go:
```diff
@@ -1,184 +1,257 @@
+	return p.client
 }
 
-// GetRecords  implements the libdns.RecordGetter interface.
```
Referring to the interface docs instead of documenting this implementation explicitly is common practice (e.g. in the std lib) and keeps it easier to maintain. Again, not really required for this lib, but I like standards...

lbrocke replied:
Could you please clarify what you mean by this? The diff looks like I removed the comment, however it's still there. Would you like me to add anything else to the comment?

Kugelschieber replied:
You don't need to provide docs here, since libdns.RecordGetter is already documented. No big deal, we keep it like this.

lbrocke replied:
Alright, thanks for clarifying. I'll mark this thread as resolved then.

Kugelschieber reviewed (dismissed):
Looks good for the most part. Did you run the tests against the new API?

Kugelschieber commented:
@lbrocke have you seen the comments?

lbrocke commented:
> @lbrocke have you seen the comments?

Sorry for the late response, I was on vacation during the last days. I pushed new changes to address your comments.

> Looks good for the most part. Did you run the tests against the new API?

Yes, I ran `provider_test.go` using multiple real zones. Additionally, I built caddy locally with my plugin changes and am able to obtain valid certificates.

lbrocke commented:
> A few remarks:
> 
> * Why not wait for the returned actions?

I'm open to doing that, I'm just not sure how Caddy or other software depending on libdns handle long running function calls. It's possible an action takes longer than 60s, which is often the default Go `context.Context` timeout.

Any opinion on this @Kugelschieber ?

> * Are we responsible for converting to punycode or is that handled elsewhere?

Good point. I couldn't find any documentation about this in libdns just now, trying it out manually showed that Caddy already converts zone names to punycode before passing it to the DNS plugin, so everything just works.

I added the punycode conversion to this plugin in commit https://github.com/libdns/hetzner/pull/11/commits/51c2dc081f16d84bdf9ddf2bb62c5442d349f6ac nonetheless, in case any other software depending on libdns doesn't do that itself.

Kugelschieber commented:
> > A few remarks:
> > 
> > * Why not wait for the returned actions?
> 
> I'm open to doing that, I'm just not sure how Caddy or other software depending on libdns handle long running function calls. It's possible an action takes longer than 60s, which is often the default Go `context.Context` timeout.
> 
> Any opinion on this @Kugelschieber ?

I also don't know how Caddy handles these cases, but I think we should wait and return an error.

Overall, lgtm if nothing else comes up!

lbrocke commented:
> > > A few remarks:
> > > 
> > > * Why not wait for the returned actions?
> > 
> > 
> > I'm open to doing that, I'm just not sure how Caddy or other software depending on libdns handle long running function calls. It's possible an action takes longer than 60s, which is often the default Go `context.Context` timeout.
> > Any opinion on this @Kugelschieber ?
> 
> I also don't know how Caddy handles these cases, but I think we should wait and return an error.
> 
> Overall, lgtm if nothing else comes up!

I added a commit to wait for actions to finish, issuing certificates still works fine with this change.

If there are no more remarks, the PR is ready to be merged. I squashed all commits in preparation for this.

Kugelschieber commented:
Awesome, thank you for contributing! :heart:
